### PR TITLE
Harvest deck/tile

### PR DESCRIPTION
**Pull request by MahChaa**
This is done and stable, if you believe there are still edge cases, share and I'll add them to the test.

### CODE REVIEW

File: NewHaven/src/Resources.h [hunk 1]
```diff
@@ -1,47 +1,69 @@
 #pragma once
+#include <cstdint>
+
+enum struct resourceTypes:std::uint_least8_t { SHEEP, STONE, WHEAT, WOOD };
```
Comment (mastrHyperion98) [inline NewHaven/src/Resources.h]: this isn't really a complaint or anything, but since you are using the standard namespace a lot why dont you just add using namespace std; after your includes? It would make it so that you don't need to write std:: before calling your functions/data types. Its up to you what you prefer. I'll continue to look through the code and double check with the assignments requirements. I should approve it somewhere around friday evening.

File: NewHaven/src/Resources.cpp [hunk 2]
```diff
@@ -111,27 +159,20 @@ HarvestDeck::~HarvestDeck()
 	delete[] this->deckContent;
 }
 
+std::uint_least8_t HarvestDeck::getDeckSize()
+{
+	return *this->deckSize;
+}
+
 HarvestTile* HarvestDeck::draw()
```
Comment (mastrHyperion98) [inline NewHaven/src/Resources.cpp:170]: shouldnt draw() be renamed drawHarvestTile() as per the assignment instructions?

File: NewHaven/src/Resources.h (hunk 1 above)
Comment (mastrHyperion98) [reply NewHaven/src/Resources.h]: Also resourceType is defined as a struct and as such the name of the data structure should start with a capital.

File: NewHaven/test/ResourcesDriver.cpp [hunk 3]
```diff
@@ -3,15 +3,39 @@
...
-    HarvestTile tile;
-    resourceTypes* content{ tile.getTileContent() };
-    std::uint_least8_t* position{ tile.getPosition() };
+    HarvestDeck deck;
+    HarvestTile* tile;
+    resourceTypes* content;
+    std::uint_least8_t deckSize;
+    std::uint_least8_t* position;
 
+    for (std::uint_fast8_t i = 0; i < 60; i++)
```
Comment (mastrHyperion98) [inline NewHaven/test/ResourcesDriver.cpp]: Okay, so I ran the code and everything compiled fine. However, the output is strange. I don't know if this is done on purpose or not but each time a card is drawn the new card is identical to the previous? Shouldn't each card in the deck be randomized? Duplicates are bound to happen but since we are working with 60 cards not all 60 of them should be the same.

File: NewHaven/src/Resources.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	delete[] this->position;
+}
+
+void HarvestTile::generateResources()
+{
+	std::uint_least8_t sheepCount{ 0 }, stoneCount{ 0 }, wheatCount{ 0 }, woodCount{ 0 };
+
+	for (std::uint_least8_t i = 0; i < 4; i++)
+	{
+		switch (std::rand() % 4 + 1)
```
Comment (mastrHyperion98) [inline NewHaven/src/Resources.cpp:45]: std::rand() is not appropriate here. The rand() function does not generate a truly random number; it actually returns the next pseudo-random value in a sequence of values ranging from 0 to RAND_MAX. This causes all 60 cards in the deck to be identical since `rand()` returns the same value throughout the lifetime of the program. You will need to import the c++ 11 random library and replace `std::rand` with `random().` If you do so this will allow your tiles to be properly randomly generated. 

However, this will cause the possibility of four different resource types appearing in a single tile. The assignment guidelines requirement
> each tile is divided into 4 squares showing 2 or 3 different Resources. 

is not met.

**Review by mastrHyperion98 — CHANGES_REQUESTED** [review]

File: NewHaven/src/Resources.h (hunk 1 above)
Comment (MahChaa) [reply NewHaven/src/Resources.h]: I prefer not using any namespaces for now just to get familiarized with the content of the standard library. I made the change to the name of the enum struct.

File: NewHaven/src/Resources.cpp (hunk 2 above)
Comment (MahChaa) [reply NewHaven/src/Resources.cpp:170]: The instructions are a bit ambiguous, but from what I understood of them is that the function in the deck class is simply draw(), DrawHarvesTile() will be a function in the player class utilizing draw() from the deck.

File: NewHaven/src/Resources.cpp (hunk 4 above)
Comment (MahChaa) [reply NewHaven/src/Resources.cpp:45]: I fixed the random number generator, the seed is now also pseudo-randomized for every tile generation. The tiles now look somewhat random.

If you look closely at the switch cases, there can not be more than 3 of the same resources on a tile, I have a counter for each resource, when the count reaches 3 and the random number generator generates the same thing again, the result is ignored and i is decremented for the loop to run again until the result is different.

File: NewHaven/src/Resources.cpp (hunk 2 above)
Comment (mastrHyperion98) [reply NewHaven/src/Resources.cpp:170]: Gotcha

File: NewHaven/src/Resources.cpp (hunk 4 above)
Comment (mastrHyperion98) [reply NewHaven/src/Resources.cpp:45]: Screenshot_20200214_131328

So I pulled the latest changes you did and ran your Driver. As seen in the above screenshot, it sometimes occur where your tile contains one of each type of resources. At least from what I understood of the instructions there should be no more than 3 different resource type on a tile. At least one resource should be duplicated.

Comment (MahChaa) [reply NewHaven/src/Resources.cpp:45]: Oh alright, I misunderstood you and I misunderstood the instructions too. I'll make 3 different resources on a tile a maximum.

Comment (mastrHyperion98) [reply NewHaven/src/Resources.cpp:45]: Yeah ^^ besides that everything seems fine. Maybe add just a few comments here and there to satisfy the "documenting" portion of the grading.

Comment (MahChaa) [reply NewHaven/src/Resources.cpp:45]: Don't worry about that, I'm trying to write up everything and then after comment the explanations to everything at the same time.

**Review by mastrHyperion98 — APPROVED** [review]
